Add a generic multimodal runner

larryliu0820 opened the pull request:
Summary:
This diff adds a generic multimodal runner for Executorch. It includes changes to the `image_prefiller.h` file, which adds a `prefill` method that takes an `Image` object and returns the next token of the LLM module after prefill. It also includes changes to the `multimodal_runner.cpp` file, which implements the `MultimodalRunner` class for multimodal input and text output LLMs. The `MultimodalRunner` class uses the `ImagePrefiller`, `TextPrefiller` classes to prefill the KV cache of the model, then uses `TextTokenGenerator` to run the autoregressive generation loop.

See diagram:

```
      ┌─────────────────┐
      │     IRunner     │
      │   <<interface>> │
      │                 │
      │ + is_loaded()   │
      │ + load()        │
      │ + generate()    │
      │ + stop()        │
      └─────────────────┘
             △
             │
             │ implements
             │
             │
             │
             │
      ┌──────┴──────────┐          ┌─────────────────┐
      │ TextLLMRunner   │          │MultimodalRunner │
      │                 │          │                 │
      │ - tokenizer_    │          │ - tokenizer_    ┼───────┐
┌─────┼ - module_       │          │ - module_       ┼─────┐ │
│ ┌───┼ - stats_        │          │ - stats_        ┼───┐ │ │
│ │ ┌─┼ - metadata_     │          │ - metadata_     ┼─┐ │ │ │
│ │ │ │ - temperature_  │          │ - pos_          │ │ │ │ │
│ │ │ └─────────────────┘          └─────────────────┘ │ │ │ │
│ │ │                                                  │ │ │ │
│ │ │                                                  │ │ │ │
│ │ │                                                  │ │ │ │
│ │ │               ┌─────────────────┐                │ │ │ │
│ │ │               │TextTokenGenerat-│                │ │ │ │
│ │ │               │or               │                │ │ │ │
│ │ │               │                 │                │ │ │ │
│ │ │               │ - tokenizer_*   │                │ │ │ │
│ │ │  consists     │ - text_decoder_ │    consists    │ │ │ │
│ │ └──────────────►│   runner_       │◄───────────────┘ │ │ │
│ │                 │ - eos_ids_      │                  │ │ │
│ │                 │ - use_kv_cache_ │                  │ │ │
│ │                 │ - stats_*       │                  │ │ │
│ │                 │                 │                  │ │ │
│ │consists         │ + generate()    │         consists │ │ │
│ │                 └────────┬────────┘                  │ │ │
│ │           ┌──────────────┴───────────────┐           │ │ │
│ │           ▼            uses              ▼           │ │ │
│ │   ┌─────────────────┐          ┌─────────────────┐   │ │ │
│ │   │TextDecoderRunner│          │MultimodalTextDe-│   │ │ │
│ │   │                 │          │coderRunner      │   │ │ │
│ │   │ - module_*      │ extends  │ - module_*      │   │ │ │
│ └──►│ - should_stop_  │◄─────────┼ - should_stop_  │◄──┘ │ │
│     │                 │          │                 │     │ │
│     │ + step()        │          │ + step()        │     │ │
│     │ + logits_to_    │          │ + logits_to_    │     │ │
│     │   token()       │          │   token()       │     │ │
│     └─────────────────┘          └─────────────────┘     │ │
│             ▲                             ▲              │ │
│             │           uses              │              │ │
│             └──────────────┬──────────────┘              │ │
│                    ┌───────┴─────────┐                   │ │
│                    │  TextPrefiller  │                   │ │
│                    │                 │                   │ │
│                    │ - text_decoder_ │                   │ │
│   consists         │   runner_       │      consists     │ │
└───────────────────►│ - use_kv_cache_ │◄──────────────────┘ │
                     │ - enable_       │                     │
                     │   parallel_     │                     │
                     │   prefill_      │                     │
                     │                 │                     │
                     │ + prefill()     │                     │
                     └─────────────────┘           consists  │
                                                             │
                                                             │
                                   ┌─────────────────┐       │
                                   │ ImagePrefiller  │       │
                                   │                 │       │
                                   │ - module_*      │       │
                                   │                 │◄──────┘
                                   │ + prefill()     │
                                   │ + logits_to_    │
                                   │   token()       │
                                   └─────────────────┘
```

Differential Revision: D79231625

msluszniak commented on extension/llm/runner/multimodal_runner.cpp:
```diff
@@ -0,0 +1,215 @@
+      get_rss_bytes() / 1024.0 / 1024.0);
+
+  // Wrap the token_callback with print function
+  std::function<void(const std::string&)> wrapped_callback =
```
Using `auto` instead of explicit `std::function` will be more efficient here.

jackzhxng reviewed (approved):
Pr not synced with internal diff so ignore these PR comments, made some comments to address internally. Otherwise looks good

larryliu0820 commented on extension/llm/runner/multimodal_input.h:
```diff
@@ -0,0 +1,167 @@
+   * @return Mutable reference to the stored text string.
+   * @throws std::bad_variant_access if this input doesn't contain text.
+   */
+  std::string& get_text() {
```
Thank you for your comments, I addressed the ones related to `multimodal_input.h` in #13310

msluszniak commented on extension/llm/runner/multimodal_input.h:
```diff
@@ -0,0 +1,167 @@
+
+// Convenience factory functions
+inline MultimodalInput make_text_input(const std::string& text) {
+  return MultimodalInput(text);
```
I guess instead of wrapping it with repetitive `MultimodalInput` you can just wrap argument with curly brackets which will call accurate constructor.

msluszniak commented on extension/llm/runner/multimodal_input.h:
```diff
@@ -0,0 +1,167 @@
+#include <variant>
+
+namespace executorch {
+namespace extension {
```
And that's general question about other files as well ;)

msluszniak commented on extension/llm/runner/multimodal_runner.cpp:
```diff
@@ -0,0 +1,215 @@
+    const MultimodalInput& input,
+    int8_t bos,
+    int8_t eos) {
+  if (input.is_text()) {
```
You can simplify this code block, so there are only if (no else if and else) since every branch returns

larryliu0820 replied:
Added image, text and else so will need `else if` for sure.

mergennachin reviewed (approved):
Very nice!

mergennachin commented:
We should make Llava runner use this new API as next step

https://github.com/pytorch/executorch/tree/main/examples/models/llava/runner

larryliu0820 committed:
Add a generic multimodal runner (#13166)

Summary:
This diff adds a generic multimodal runner for Executorch. It includes changes to the `image_prefiller.h` file, which adds a `prefill` method that takes an `Image` object and returns the next token of the LLM module after prefill. It also includes changes to the `multimodal_runner.cpp` file, which implements the `MultimodalRunner` class for multimodal input and text output LLMs. The `MultimodalRunner` class uses the `ImagePrefiller`, `TextPrefiller` classes to prefill the KV cache of the model, then uses `TextTokenGenerator` to run the autoregressive generation loop.

See diagram:

```
      ┌─────────────────┐
      │     IRunner     │
      │   <<interface>> │
      │                 │
      │ + is_loaded()   │
      │ + load()        │
      │ + generate()    │
      │ + stop()        │
      └─────────────────┘
             △
             │
             │ implements
             │
             │
             │
             │
      ┌──────┴──────────┐          ┌─────────────────┐
      │ TextLLMRunner   │          │MultimodalRunner │
      │                 │          │                 │
      │ - tokenizer_    │          │ - tokenizer_    ┼───────┐
┌─────┼ - module_       │          │ - module_       ┼─────┐ │
│ ┌───┼ - stats_        │          │ - stats_        ┼───┐ │ │
│ │ ┌─┼ - metadata_     │          │ - metadata_     ┼─┐ │ │ │
│ │ │ │ - temperature_  │          │ - pos_          │ │ │ │ │
│ │ │ └─────────────────┘          └─────────────────┘ │ │ │ │
│ │ │                                                  │ │ │ │
│ │ │                                                  │ │ │ │
│ │ │                                                  │ │ │ │
│ │ │               ┌─────────────────┐                │ │ │ │
│ │ │               │TextTokenGenerat-│                │ │ │ │
│ │ │               │or               │                │ │ │ │
│ │ │               │                 │                │ │ │ │
│ │ │               │ - tokenizer_*   │                │ │ │ │
│ │ │  consists     │ - text_decoder_ │    consists    │ │ │ │
│ │ └──────────────►│   runner_       │◄───────────────┘ │ │ │
│ │                 │ - eos_ids_      │                  │ │ │
│ │                 │ - use_kv_cache_ │                  │ │ │
│ │                 │ - stats_*       │                  │ │ │
│ │                 │                 │                  │ │ │
│ │consists         │ + generate()    │         consists │ │ │
│ │                 └────────┬────────┘                  │ │ │
│ │           ┌──────────────┴───────────────┐           │ │ │
│ │           ▼            uses              ▼           │ │ │
│ │   ┌─────────────────┐          ┌─────────────────┐   │ │ │
│ │   │TextDecoderRunner│          │MultimodalTextDe-│   │ │ │
│ │   │                 │          │coderRunner      │   │ │ │
│ │   │ - module_*      │ extends  │ - module_*      │   │ │ │
│ └──►│ - should_stop_  │◄─────────┼ - should_stop_  │◄──┘ │ │
│     │                 │          │                 │     │ │
│     │ + step()        │          │ + step()        │     │ │
│     │ + logits_to_    │          │ + logits_to_    │     │ │
│     │   token()       │          │   token()       │     │ │
│     └─────────────────┘          └─────────────────┘     │ │
│             ▲                             ▲              │ │
│             │           uses              │              │ │
│             └──────────────┬──────────────┘              │ │
│                    ┌───────┴─────────┐                   │ │
│                    │  TextPrefiller  │                   │ │
│                    │                 │                   │ │
│                    │ - text_decoder_ │                   │ │
│   consists         │   runner_       │      consists     │ │
└───────────────────►│ - use_kv_cache_ │◄──────────────────┘ │
                     │ - enable_       │                     │
                     │   parallel_     │                     │
                     │   prefill_      │                     │
                     │                 │                     │
                     │ + prefill()     │                     │
                     └─────────────────┘           consists  │
                                                             │
                                                             │
                                   ┌─────────────────┐       │
                                   │ ImagePrefiller  │       │
                                   │                 │       │
                                   │ - module_*      │       │
                                   │                 │◄──────┘
                                   │ + prefill()     │
                                   │ + logits_to_    │
                                   │   token()       │
                                   └─────────────────┘
```

Pull Request resolved: https://github.com/pytorch/executorch/pull/13166

Test Plan:
Imported from GitHub, without a `Test Plan:` line.

Rollback Plan:

Reviewed By: mergennachin

Differential Revision: D79231625

Pulled By: larryliu0820
